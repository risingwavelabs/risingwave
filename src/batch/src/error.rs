// Copyright 2025 RisingWave Labs
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use std::sync::Arc;

pub use anyhow::anyhow;
use mysql_async::Error as MySqlError;
use parquet::errors::ParquetError;
use risingwave_common::array::ArrayError;
use risingwave_common::error::{BoxedError, IcebergError, def_anyhow_newtype, def_anyhow_variant};
use risingwave_common::util::value_encoding::error::ValueEncodingError;
use risingwave_connector::error::ConnectorError;
use risingwave_dml::error::DmlError;
use risingwave_expr::ExprError;
use risingwave_pb::PbFieldNotFound;
use risingwave_rpc_client::error::{RpcError, ToTonicStatus};
use risingwave_storage::error::StorageError;
use thiserror::Error;
use thiserror_ext::Construct;
use tokio_postgres::Error as PostgresError;
use tonic::Status;

use crate::worker_manager::worker_node_manager::FragmentId;

pub type Result<T> = std::result::Result<T, BatchError>;
/// Batch result with shared error.
pub type SharedResult<T> = std::result::Result<T, Arc<BatchError>>;

pub trait Error = std::error::Error + Send + Sync + 'static;

#[derive(Error, Debug, Construct)]
pub enum BatchError {
    #[error("Storage error: {0}")]
    Storage(
        #[backtrace]
        #[from]
        StorageError,
    ),

    #[error("Array error: {0}")]
    Array(
        #[from]
        #[backtrace]
        ArrayError,
    ),

    #[error("Expr error: {0}")]
    Expr(
        #[from]
        #[backtrace]
        ExprError,
    ),

    #[error("Serialize/deserialize error: {0}")]
    Serde(
        #[source]
        #[backtrace]
        BoxedError,
    ),

    #[error("Failed to send result to channel")]
    SenderError,

    #[error(transparent)]
    Internal(
        #[from]
        #[backtrace]
        anyhow::Error,
    ),

    #[error("Task aborted: {0}")]
    Aborted(String),

    #[error(transparent)]
    PbFieldNotFound(#[from] PbFieldNotFound),

    #[error(transparent)]
    RpcError(
        #[from]
        #[backtrace]
        RpcError,
    ),

    #[error("Connector error: {0}")]
    Connector(
        #[source]
        #[backtrace]
        BoxedError,
    ),

    #[error("Failed to read from system table: {0}")]
    SystemTable(
        #[from]
        #[backtrace]
        BoxedError,
    ),

    #[error(transparent)]
    Dml(
        #[from]
        #[backtrace]
        DmlError,
    ),

    #[error("External system error: {0}")]
    ExternalSystemError(
        #[from]
        #[backtrace]
        BatchExternalSystemError,
    ),

    // Make the ref-counted type to be a variant for easier code structuring.
    // TODO(error-handling): replace with `thiserror_ext::Arc`
    #[error(transparent)]
    Shared(
        #[from]
        #[backtrace]
        Arc<Self>,
    ),

    #[error("Empty workers found")]
    EmptyWorkerNodes,

    #[error("Serving vnode mapping not found for fragment {0}")]
    ServingVnodeMappingNotFound(FragmentId),

    #[error("Streaming vnode mapping not found for fragment {0}")]
    StreamingVnodeMappingNotFound(FragmentId),

    #[error("Not enough memory to run this query, batch memory limit is {0} bytes")]
    OutOfMemory(u64),

    #[error("Failed to spill out to disk")]
    Spill(
        #[from]
        #[backtrace]
        opendal::Error,
    ),

    #[error("Failed to execute time travel query")]
    TimeTravel(
        #[source]
        #[backtrace]
        anyhow::Error,
    ),
}

// Serialize/deserialize error.
impl From<memcomparable::Error> for BatchError {
    fn from(m: memcomparable::Error) -> Self {
        Self::Serde(m.into())
    }
}
impl From<ValueEncodingError> for BatchError {
    fn from(e: ValueEncodingError) -> Self {
        Self::Serde(e.into())
    }
}

impl<'a> From<&'a BatchError> for Status {
    fn from(err: &'a BatchError) -> Self {
        err.to_status(tonic::Code::Internal, "batch")
    }
}

impl From<BatchError> for Status {
    fn from(err: BatchError) -> Self {
        Self::from(&err)
    }
}

impl From<ConnectorError> for BatchError {
    fn from(value: ConnectorError) -> Self {
        Self::Connector(value.into())
    }
}

// Define a external system error
def_anyhow_variant! {
    pub BatchExternalSystemError,
    BatchError ExternalSystemError,

    PostgresError => "Postgres error",
    IcebergError => "Iceberg error",
    ParquetError => "Parquet error",
    MySqlError => "MySQL error",
}

#[expect(clippy::disallowed_types)]
impl From<iceberg::Error> for BatchExternalSystemError {
    fn from(value: iceberg::Error) -> Self {
        risingwave_common::error::IcebergError::from(value).into()
    }
}
