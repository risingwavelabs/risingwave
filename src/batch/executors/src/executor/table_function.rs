// Copyright 2025 RisingWave Labs
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use futures_async_stream::try_stream;
use risingwave_common::array::{ArrayImpl, DataChunk};
use risingwave_common::catalog::{Field, Schema};
use risingwave_common::types::DataType;
use risingwave_expr::table_function::{BoxedTableFunction, build_from_prost, check_error};
use risingwave_pb::batch_plan::plan_node::NodeBody;

use super::{BoxedExecutor, BoxedExecutorBuilder};
use crate::error::{BatchError, Result};
use crate::executor::{BoxedDataChunkStream, Executor, ExecutorBuilder};

pub struct TableFunctionExecutor {
    schema: Schema,
    identity: String,
    table_function: BoxedTableFunction,
    #[expect(dead_code)]
    chunk_size: usize,
}

impl Executor for TableFunctionExecutor {
    fn schema(&self) -> &Schema {
        &self.schema
    }

    fn identity(&self) -> &str {
        &self.identity
    }

    fn execute(self: Box<Self>) -> BoxedDataChunkStream {
        self.do_execute()
    }
}

impl TableFunctionExecutor {
    #[try_stream(boxed, ok = DataChunk, error = BatchError)]
    async fn do_execute(self: Box<Self>) {
        let dummy_chunk = DataChunk::new_dummy(1);

        #[for_await]
        for chunk in self.table_function.eval(&dummy_chunk).await {
            let chunk = chunk?;
            check_error(&chunk)?;
            // remove the first column and expand the second column if its data type is struct
            yield match chunk.column_at(1).as_ref() {
                ArrayImpl::Struct(struct_array) => struct_array.into(),
                _ => chunk.split_column_at(1).1,
            };
        }
    }
}

pub struct TableFunctionExecutorBuilder {}

impl TableFunctionExecutorBuilder {}

impl BoxedExecutorBuilder for TableFunctionExecutorBuilder {
    async fn new_boxed_executor(
        source: &ExecutorBuilder<'_>,
        inputs: Vec<BoxedExecutor>,
    ) -> Result<BoxedExecutor> {
        ensure!(
            inputs.is_empty(),
            "GenerateSeriesExecutor should not have child!"
        );
        let node = try_match_expand!(
            source.plan_node().get_node_body().unwrap(),
            NodeBody::TableFunction
        )?;

        let identity = source.plan_node().get_identity().clone();

        let chunk_size = source.context().get_config().developer.chunk_size;

        let table_function = build_from_prost(node.table_function.as_ref().unwrap(), chunk_size)?;

        let schema = if let DataType::Struct(fields) = table_function.return_type() {
            (&fields).into()
        } else {
            Schema {
                // TODO: should be named
                fields: vec![Field::unnamed(table_function.return_type())],
            }
        };

        Ok(Box::new(TableFunctionExecutor {
            schema,
            identity,
            table_function,
            chunk_size,
        }))
    }
}
