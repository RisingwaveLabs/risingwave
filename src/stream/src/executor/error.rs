// Copyright 2022 Singularity Data
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use std::backtrace::Backtrace;

use risingwave_common::array::ArrayError;
use risingwave_common::error::{BoxedError, ErrorCode, RwError, TrackingIssue};
use risingwave_expr::ExprError;
use risingwave_storage::error::StorageError;
use thiserror::Error;

use super::Barrier;

#[derive(Error, Debug)]
enum StreamExecutorErrorInner {
    #[error("Storage error: {0}")]
    Storage(
        #[backtrace]
        #[source]
        StorageError,
    ),

    #[error("Invalid argument: {0}")]
    InvalidArgument(String),

    #[error("Executor v1 error: {0}")]
    ExecutorV1(RwError),

    #[error("Chunk operation error: {0}")]
    EvalError(BoxedError),

    // TODO: remove this after state table is fully used
    #[error("Serialize/deserialize error: {0}")]
    SerdeError(BoxedError),

    #[error("Input error: {0}")]
    InputError(RwError),

    #[error("TopN state error: {0}")]
    TopNStateError(RwError),

    #[error("Hash join error: {0}")]
    HashJoinError(RwError),

    #[error("Source error: {0}")]
    SourceError(RwError),

    #[error("Channel `{0}` closed")]
    ChannelClosed(String),

    #[error("Failed to align barrier: expected {0:?} but got {1:?}")]
    AlignBarrier(Box<Barrier>, Box<Barrier>),

    #[error("Feature is not yet implemented: {0}, {1}")]
    NotImplemented(String, TrackingIssue),

    #[error(transparent)]
    Internal(anyhow::Error),
}

impl StreamExecutorError {
    pub fn storage(error: impl Into<StorageError>) -> Self {
        StreamExecutorErrorInner::Storage(error.into()).into()
    }

    pub fn executor_v1(error: impl Into<RwError>) -> Self {
        StreamExecutorErrorInner::ExecutorV1(error.into()).into()
    }

    pub fn eval_error(error: impl std::error::Error + Send + Sync) -> Self {
        StreamExecutorErrorInner::EvalError(error.into()).into()
    }

    pub fn serde_error(error: impl std::error::Error + Send + Sync) -> Self {
        StreamExecutorErrorInner::SerdeError(error.into()).into()
    }

    pub fn input_error(error: impl Into<RwError>) -> Self {
        StreamExecutorErrorInner::InputError(error.into()).into()
    }

    pub fn top_n_state_error(error: impl Into<RwError>) -> Self {
        StreamExecutorErrorInner::TopNStateError(error.into()).into()
    }

    pub fn hash_join_error(error: impl Into<RwError>) -> Self {
        StreamExecutorErrorInner::HashJoinError(error.into()).into()
    }

    pub fn source_error(error: impl Into<RwError>) -> Self {
        StreamExecutorErrorInner::SourceError(error.into()).into()
    }

    pub fn channel_closed(name: impl Into<String>) -> Self {
        StreamExecutorErrorInner::ChannelClosed(name.into()).into()
    }

    pub fn align_barrier(expected: Barrier, received: Barrier) -> Self {
        StreamExecutorErrorInner::AlignBarrier(expected.into(), received.into()).into()
    }

    pub fn invalid_argument(error: impl Into<String>) -> Self {
        StreamExecutorErrorInner::InvalidArgument(error.into()).into()
    }

    pub fn not_implemented(error: impl Into<String>, issue: impl Into<TrackingIssue>) -> Self {
        StreamExecutorErrorInner::NotImplemented(error.into(), issue.into()).into()
    }
}

#[derive(Error)]
#[error("{inner}")]
pub struct StreamExecutorError {
    #[from]
    inner: StreamExecutorErrorInner,
    backtrace: Backtrace,
}

impl std::fmt::Debug for StreamExecutorError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        use std::error::Error;

        write!(f, "{}", self.inner)?;
        writeln!(f)?;
        if let Some(backtrace) = self.inner.backtrace() {
            write!(f, "  backtrace of inner error:\n{}", backtrace)?;
        } else {
            write!(
                f,
                "  backtrace of `StreamExecutorError`:\n{}",
                self.backtrace
            )?;
        }
        Ok(())
    }
}

impl From<StorageError> for StreamExecutorError {
    fn from(s: StorageError) -> Self {
        Self::storage(s)
    }
}

impl From<ArrayError> for StreamExecutorError {
    fn from(a: ArrayError) -> Self {
        Self::eval_error(a)
    }
}

impl From<ExprError> for StreamExecutorError {
    fn from(e: ExprError) -> Self {
        Self::eval_error(e)
    }
}

impl From<anyhow::Error> for StreamExecutorError {
    fn from(a: anyhow::Error) -> Self {
        StreamExecutorErrorInner::Internal(a).into()
    }
}

impl From<memcomparable::Error> for StreamExecutorError {
    fn from(m: memcomparable::Error) -> Self {
        Self::serde_error(m)
    }
}

/// Always convert [`StreamExecutorError`] to stream error variant of [`RwError`].
impl From<StreamExecutorError> for RwError {
    fn from(h: StreamExecutorError) -> Self {
        ErrorCode::StreamError(h.into()).into()
    }
}

pub type StreamExecutorResult<T> = std::result::Result<T, StreamExecutorError>;

#[cfg(test)]
mod tests {
    use super::*;

    fn func_return_error() -> StreamExecutorResult<()> {
        Err(ErrorCode::InternalError("test_error".into())).map_err(StreamExecutorError::executor_v1)
    }

    #[test]
    #[should_panic]
    #[ignore]
    fn executor_error_ui_test_1() {
        // For this test, ensure that we have only one backtrace from error when panic.
        func_return_error().unwrap();
    }

    #[test]
    #[ignore]
    fn executor_error_ui_test_2() {
        // For this test, ensure that we have only one backtrace from error when panic.
        func_return_error().map_err(|e| println!("{:?}", e)).ok();
    }
}
