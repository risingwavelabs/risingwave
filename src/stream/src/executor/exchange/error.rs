// Copyright 2025 RisingWave Labs
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use risingwave_rpc_client::error::TonicStatusWrapper;

use crate::task::ActorId;

/// The error type for the exchange channel closed unexpectedly.
///
/// In most cases, this error happens when the upstream or downstream actor
/// exits or panics on other errors, or the network connection is broken.
/// Therefore, this error is usually not the root case of the failure in the
/// streaming graph.
#[derive(Debug)]
pub struct ExchangeChannelClosed {
    message: String,

    /// `Some` if there is a gRPC error from the remote actor.
    source: Option<TonicStatusWrapper>,
}

impl std::fmt::Display for ExchangeChannelClosed {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.message)
    }
}

impl std::error::Error for ExchangeChannelClosed {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        self.source.as_ref().map(|s| s as _)
    }

    fn provide<'a>(&'a self, request: &mut std::error::Request<'a>) {
        use std::backtrace::Backtrace;

        // Always provide a fake disabled backtrace, so that the upper layer will
        // not capture any other backtraces or include the backtrace in the error
        // log.
        //
        // Otherwise, when an actor exits on a significant error, all connected
        // actor will then exit with the `ExchangeChannelClosed` error, resulting
        // in a very noisy log with flood of useless backtraces.
        static DISABLED_BACKTRACE: Backtrace = Backtrace::disabled();
        request.provide_ref::<Backtrace>(&DISABLED_BACKTRACE);

        if let Some(source) = &self.source {
            source.provide(request);
        }
    }
}

impl ExchangeChannelClosed {
    /// Creates a new error indicating that the exchange channel from the local
    /// upstream actor is closed unexpectedly.
    pub fn local_input(upstream: ActorId) -> Self {
        Self {
            message: format!(
                "exchange channel from local upstream actor {upstream} closed unexpectedly"
            ),
            source: None,
        }
    }

    /// Creates a new error indicating that the exchange channel from the remote
    /// upstream actor is closed unexpectedly, with an optional gRPC error as the cause.
    pub fn remote_input(upstream: ActorId, source: Option<tonic::Status>) -> Self {
        Self {
            message: format!(
                "exchange channel from remote upstream actor {upstream} closed unexpectedly"
            ),
            source: source.map(Into::into),
        }
    }

    /// Creates a new error indicating that the exchange channel to the downstream
    /// actor is closed unexpectedly.
    pub fn output(downstream: ActorId) -> Self {
        Self {
            message: format!(
                "exchange channel to downstream actor {downstream} closed unexpectedly"
            ),
            source: None,
        }
    }
}
