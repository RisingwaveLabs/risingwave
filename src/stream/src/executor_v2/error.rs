// Copyright 2022 Singularity Data
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use std::backtrace::Backtrace;

use risingwave_common::error::{ErrorCode, RwError};
use risingwave_storage::error::StorageError;
use thiserror::Error;

use super::Barrier;

#[derive(Error, Debug)]
pub enum StreamExecutorError {
    #[error("Storage error: {0}")]
    Storage(
        #[backtrace]
        #[source]
        StorageError,
    ),

    #[error("Invalid argument: {0}")]
    InvalidArgument(String),

    #[error("Executor v1 error: {0}")]
    ExecutorV1(RwError),

    #[error("Chunk operation error: {0}")]
    EvalError(RwError),

    #[error("Aggregate state error: {0}")]
    AggStateError(RwError),

    #[error("Input error: {0}")]
    InputError(RwError),

    #[error("TopN state error: {0}")]
    TopNStateError(RwError),

    #[error("Channel `{0}` closed")]
    ChannelClosed(String),

    #[error("Failed to align barrier: expected {0:?} but got {1:?}")]
    AlignBarrier(Box<Barrier>, Box<Barrier>),
}

impl StreamExecutorError {
    pub fn storage(error: impl Into<StorageError>) -> TracedStreamExecutorError {
        Self::Storage(error.into()).into()
    }

    pub fn executor_v1(error: impl Into<RwError>) -> TracedStreamExecutorError {
        Self::ExecutorV1(error.into()).into()
    }

    pub fn eval_error(error: impl Into<RwError>) -> TracedStreamExecutorError {
        Self::EvalError(error.into()).into()
    }

    pub fn agg_state_error(error: impl Into<RwError>) -> TracedStreamExecutorError {
        Self::AggStateError(error.into()).into()
    }

    pub fn input_error(error: impl Into<RwError>) -> TracedStreamExecutorError {
        Self::InputError(error.into()).into()
    }

    pub fn top_n_state_error(error: impl Into<RwError>) -> TracedStreamExecutorError {
        Self::TopNStateError(error.into()).into()
    }

    pub fn channel_closed(name: impl Into<String>) -> TracedStreamExecutorError {
        Self::ChannelClosed(name.into()).into()
    }

    pub fn align_barrier(expected: Barrier, received: Barrier) -> TracedStreamExecutorError {
        Self::AlignBarrier(expected.into(), received.into()).into()
    }
}

#[derive(Error)]
#[error("{source}")]
pub struct TracedStreamExecutorError {
    #[from]
    source: StreamExecutorError,
    backtrace: Backtrace,
}

impl std::fmt::Debug for TracedStreamExecutorError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        use std::error::Error;

        write!(f, "{}", self.source)?;
        writeln!(f)?;
        if let Some(backtrace) = self.source.backtrace() {
            write!(f, "  backtrace of inner error:\n{}", backtrace)?;
        } else {
            write!(
                f,
                "  backtrace of `TracedStreamExecutorError`:\n{}",
                self.backtrace
            )?;
        }
        Ok(())
    }
}

impl From<StorageError> for TracedStreamExecutorError {
    fn from(s: StorageError) -> Self {
        StreamExecutorError::storage(s)
    }
}

/// Always convert [`TracedStreamExecutorError`] to stream error of `RwResult`.
impl From<TracedStreamExecutorError> for RwError {
    fn from(h: TracedStreamExecutorError) -> Self {
        ErrorCode::StreamError(h.into()).into()
    }
}

pub type StreamExecutorResult<T> = std::result::Result<T, TracedStreamExecutorError>;

#[cfg(test)]
mod tests {
    use super::*;

    fn func_return_error() -> StreamExecutorResult<()> {
        Err(ErrorCode::InternalError("test_error".into())).map_err(StreamExecutorError::executor_v1)
    }

    #[test]
    #[should_panic]
    #[ignore]
    fn executor_error_ui_test_1() {
        // For this test, ensure that we have only one backtrace from error when panic.
        func_return_error().unwrap();
    }

    #[test]
    #[ignore]
    fn executor_error_ui_test_2() {
        // For this test, ensure that we have only one backtrace from error when panic.
        func_return_error().map_err(|e| println!("{:?}", e)).ok();
    }
}
