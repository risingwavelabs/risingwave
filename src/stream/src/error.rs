// Copyright 2025 RisingWave Labs
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use risingwave_common::array::ArrayError;
use risingwave_common::secret::SecretError;
use risingwave_connector::error::ConnectorError;
use risingwave_expr::ExprError;
use risingwave_pb::PbFieldNotFound;
use risingwave_rpc_client::error::ToTonicStatus;
use risingwave_storage::error::StorageError;

use crate::executor::exchange::error::ExchangeChannelClosed;
use crate::executor::{Barrier, StreamExecutorError};
use crate::task::ActorId;

/// A specialized Result type for streaming tasks.
pub type StreamResult<T> = std::result::Result<T, StreamError>;

/// The error type for streaming tasks.
#[derive(
    thiserror::Error,
    thiserror_ext::ReportDebug,
    thiserror_ext::Arc,
    thiserror_ext::ContextInto,
    thiserror_ext::Construct,
)]
#[thiserror_ext(newtype(name = StreamError, backtrace))]
pub enum ErrorKind {
    #[error("Storage error: {0}")]
    Storage(
        #[backtrace]
        #[from]
        StorageError,
    ),

    #[error("Expression error: {0}")]
    Expression(
        #[from]
        #[backtrace]
        ExprError,
    ),

    #[error("Array/Chunk error: {0}")]
    Array(
        #[from]
        #[backtrace]
        ArrayError,
    ),

    #[error("Executor error: {0}")]
    Executor(
        #[from]
        #[backtrace]
        StreamExecutorError,
    ),

    #[error("Actor {actor_id} exited unexpectedly: {source}")]
    UnexpectedExit {
        actor_id: ActorId,
        #[backtrace]
        source: StreamError,
    },

    #[error("Failed to send barrier with epoch {epoch} to actor {actor_id}: {reason}", epoch = .barrier.epoch.curr)]
    BarrierSend {
        barrier: Barrier,
        actor_id: ActorId,
        reason: &'static str,
    },

    #[error("Secret error: {0}")]
    Secret(
        #[from]
        #[backtrace]
        SecretError,
    ),

    #[error(transparent)]
    Uncategorized(
        #[from]
        #[backtrace]
        anyhow::Error,
    ),
}

impl From<PbFieldNotFound> for StreamError {
    fn from(err: PbFieldNotFound) -> Self {
        Self::from(anyhow::anyhow!(
            "Failed to decode prost: field not found `{}`",
            err.0
        ))
    }
}

impl From<ConnectorError> for StreamError {
    fn from(err: ConnectorError) -> Self {
        StreamExecutorError::from(err).into()
    }
}

impl From<ExchangeChannelClosed> for StreamError {
    fn from(err: ExchangeChannelClosed) -> Self {
        StreamExecutorError::from(err).into()
    }
}

impl From<StreamError> for tonic::Status {
    fn from(error: StreamError) -> Self {
        error.to_status(tonic::Code::Internal, "stream")
    }
}

static_assertions::const_assert_eq!(std::mem::size_of::<StreamError>(), 8);
