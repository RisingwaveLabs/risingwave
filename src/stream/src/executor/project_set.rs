// Copyright 2022 Singularity Data
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use std::fmt::{Debug, Formatter};
use std::sync::Arc;

use futures::StreamExt;
use futures_async_stream::try_stream;
use risingwave_common::array::StreamChunk;
use risingwave_common::catalog::{Field, Schema};
use risingwave_expr::table_function::ProjectSetSelectItem;

use super::error::StreamExecutorError;
use super::{BoxedExecutor, Executor, ExecutorInfo, Message, PkIndices, PkIndicesRef};

impl ProjectSetExecutor {
    pub fn new(
        input: Box<dyn Executor>,
        pk_indices: PkIndices,
        select_list: Vec<ProjectSetSelectItem>,
        executor_id: u64,
    ) -> Self {
        let fields = select_list
            .iter()
            .map(|expr| Field::unnamed(expr.return_type()))
            .collect();

        let info = ExecutorInfo {
            schema: Schema { fields },
            pk_indices,
            identity: format!("ProjectSet {:X}", executor_id),
        };
        Self {
            input,
            info,
            select_list,
        }
    }
}

/// `ProjectSetExecutor` projects data with the `expr`. The `expr` takes a chunk of data,
/// and returns a new data chunk. And then, `ProjectSetExecutor` will insert, delete
/// or update element into next operator according to the result of the expression.
pub struct ProjectSetExecutor {
    input: BoxedExecutor,
    info: ExecutorInfo,
    /// Expressions of the current project_setion.
    select_list: Vec<ProjectSetSelectItem>,
}

impl Debug for ProjectSetExecutor {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("ProjectSetExecutor")
            .field("exprs", &self.select_list)
            .finish()
    }
}

impl Executor for ProjectSetExecutor {
    fn execute(self: Box<Self>) -> super::BoxedMessageStream {
        self.execute_inner().boxed()
    }

    fn schema(&self) -> &Schema {
        &self.info.schema
    }

    fn pk_indices(&self) -> PkIndicesRef {
        &self.info.pk_indices
    }

    fn identity(&self) -> &str {
        &self.info.identity
    }
}

impl ProjectSetExecutor {
    #[try_stream(ok = Message, error = StreamExecutorError)]
    async fn execute_inner(self) {
        let data_types = self.schema().data_types();
        let input = self.input.execute();
        let select_list = Arc::new(self.select_list);

        #[for_await]
        for msg in input {
            let msg = msg?;
            match msg {
                Message::Chunk(chunk) => {
                    let chunk = chunk.compact()?;

                    let (data_chunk, ops) = chunk.into_parts();

                    #[for_await]
                    for (i, ret) in
                        ProjectSetSelectItem::execute(&select_list, &data_types, &data_chunk)
                            .enumerate()
                    {
                        let ret = ret.map_err(StreamExecutorError::eval_error)?;
                        let new_chunk =
                            StreamChunk::from_parts(vec![ops[i]; ret.cardinality()], ret);
                        yield Message::Chunk(new_chunk)
                    }
                }
                m => yield m,
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use futures::StreamExt;
    use risingwave_common::array::stream_chunk::StreamChunkTestExt;
    use risingwave_common::array::StreamChunk;
    use risingwave_common::catalog::{Field, Schema};
    use risingwave_common::types::DataType;
    use risingwave_expr::expr::expr_binary_nonnull::new_binary_expr;
    use risingwave_expr::expr::{Expression, InputRefExpression, LiteralExpression};
    use risingwave_expr::table_function::repeat_tf;
    use risingwave_pb::expr::expr_node::Type;

    use super::super::test_utils::MockSource;
    use super::super::*;
    use super::*;

    #[tokio::test]
    async fn test_project_set() {
        let chunk1 = StreamChunk::from_pretty(
            " I I
            + 1 4
            + 2 5
            + 3 6",
        );
        let chunk2 = StreamChunk::from_pretty(
            " I I
            + 7 8
            - 3 6",
        );
        let schema = Schema {
            fields: vec![
                Field::unnamed(DataType::Int64),
                Field::unnamed(DataType::Int64),
            ],
        };
        let source = MockSource::with_chunks(schema, PkIndices::new(), vec![chunk1, chunk2]);

        let left_expr = InputRefExpression::new(DataType::Int64, 0);
        let right_expr = InputRefExpression::new(DataType::Int64, 1);
        let test_expr = new_binary_expr(
            Type::Add,
            DataType::Int64,
            Box::new(left_expr),
            Box::new(right_expr),
        );
        let tf1 = repeat_tf(
            LiteralExpression::new(DataType::Int32, Some(1_i32.into())).boxed(),
            1,
        );
        let tf2 = repeat_tf(
            LiteralExpression::new(DataType::Int32, Some(2_i32.into())).boxed(),
            2,
        );

        let project_set = Box::new(ProjectSetExecutor::new(
            Box::new(source),
            vec![],
            vec![test_expr.into(), tf1.into(), tf2.into()],
            1,
        ));

        let expected = vec![
            StreamChunk::from_pretty(
                " I I i i
                + 0 5 1 2
                + 1 5 . 2",
            ),
            StreamChunk::from_pretty(
                " I I i i
                + 0 7 1 2
                + 1 7 . 2",
            ),
            StreamChunk::from_pretty(
                " I I i i
                + 0 9 1 2
                + 1 9 . 2",
            ),
            StreamChunk::from_pretty(
                " I I  i i
                + 0 15 1 2
                + 1 15 . 2",
            ),
            StreamChunk::from_pretty(
                " I I i i
                - 0 9 1 2
                - 1 9 . 2",
            ),
        ];

        let mut project_set = project_set.execute();

        for expected in expected {
            let msg = project_set.next().await.unwrap().unwrap();
            let chunk = msg.as_chunk().unwrap();
            assert_eq!(*chunk, expected);
        }
        assert!(project_set.next().await.unwrap().unwrap().is_stop());
    }
}
