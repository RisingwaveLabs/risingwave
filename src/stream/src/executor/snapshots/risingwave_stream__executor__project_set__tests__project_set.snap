---
source: src/stream/src/executor/project_set.rs
description: "($0+$1), repeat(1,1), repeat(2,2)"
---
# This result can be automatically updated. See `TODO:` for more information.
- input: !chunk |-
    +---+---+---+
    | + | 1 | 4 |
    | + | 2 | 5 |
    | + | 3 | 6 |
    +---+---+---+
  output:
  - !chunk |-
    +---+---+---+---+---+
    | + | 0 | 5 | 1 | 2 |
    | + | 1 | 5 |   | 2 |
    | + | 0 | 7 | 1 | 2 |
    | + | 1 | 7 |   | 2 |
    | + | 0 | 9 | 1 | 2 |
    | + | 1 | 9 |   | 2 |
    +---+---+---+---+---+
- input: !chunk |-
    +---+---+---+
    | + | 7 | 8 |
    | - | 3 | 6 |
    +---+---+---+
  output:
  - !chunk |-
    +---+---+----+---+---+
    | + | 0 | 15 | 1 | 2 |
    | + | 1 | 15 |   | 2 |
    | - | 0 | 9  | 1 | 2 |
    | - | 1 | 9  |   | 2 |
    +---+---+----+---+---+

